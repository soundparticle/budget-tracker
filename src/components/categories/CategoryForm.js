import React, { Component } from 'react';
import PropTypes from 'prop-types';

class CategoryForm extends Component {
  state = {
    editing: false,
    key: null,
    name: '',
    budget: '',    
    expenses: [],
    category: null,
  };

  static propTypes = {
    category: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };
  
  componentDidMount() {
    const { category } = this.props;
    if(!category) return;

    this.setState(category);
  }
  
  handleSubmit = (event) => {
    event.preventDefault();
    const { name, budget, key } = this.state;
    const category = { name, budget };
    if(key) category.key = key;

    this.props.onComplete(category);
    this.setState({ name: '', budget: '' });
  };
  
  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() { 
    const { key, name, budget } = this.state;
    const { onCancel } = this.props;

    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Name:
          <input name="name" value={name} onChange={this.handleChange}/>
        </label>
        <label>
          Budget:
          <input name="budget" value={budget} onChange={this.handleChange}/>
        </label>
        
        <div>
          <span>
            <button type="submit">{ key ? 'Update' : 'Add' }</button>
            {key && <button type="button" onClick={onCancel}>Cancel</button>}
          </span>
        </div>
      </form>
    );
  }
}

export default CategoryForm;
