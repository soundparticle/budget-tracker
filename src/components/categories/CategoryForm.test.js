import CategoryForm from './CategoryForm';
import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

describe('Category form', () => {
  
  it('renders add if no category prop', () => {
    const handleComplete = jest.fn();
    const promise = Promise.resolve();
    handleComplete.mockReturnValueOnce(promise);

    const wrapper = mount(<CategoryForm onComplete={handleComplete}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();

    const category = {
      name: 'shopping',
      budget: '400'
    };

    wrapper.find('input[name="name"]').simulate('change', {
      target: {
        name: 'name',
        value: category.name
      }
    });

    wrapper.find('input[name="budget"]').simulate('change', {
      target: {
        name: 'budget',
        value: category.budget
      }
    });

    wrapper.find('button').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual(category);

    expect(toJSON(wrapper)).toMatchSnapshot();

    return promise
      .then(() => {
        wrapper.update();
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
  });
});