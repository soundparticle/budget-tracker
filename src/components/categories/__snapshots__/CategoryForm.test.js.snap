// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category form renders add if no category prop 1`] = `
<CategoryForm
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name:
      <input
        name="name"
        onChange={[Function]}
        value=""
      />
    </label>
    <label>
      Budget:
      <input
        name="budget"
        onChange={[Function]}
        value=""
      />
    </label>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</CategoryForm>
`;

exports[`Category form renders add if no category prop 2`] = `
<CategoryForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "budget": "400",
            "name": "shopping",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name:
      <input
        name="name"
        onChange={[Function]}
        value=""
      />
    </label>
    <label>
      Budget:
      <input
        name="budget"
        onChange={[Function]}
        value=""
      />
    </label>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</CategoryForm>
`;

exports[`Category form renders add if no category prop 3`] = `
<CategoryForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "budget": "400",
            "name": "shopping",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name:
      <input
        name="name"
        onChange={[Function]}
        value=""
      />
    </label>
    <label>
      Budget:
      <input
        name="budget"
        onChange={[Function]}
        value=""
      />
    </label>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</CategoryForm>
`;
